mlflow/utils/conda.py

```diff
@@ -126,36 +131,55 @@ def get_or_create_conda_env(conda_env_path, env_id=None):
             )
         )
 
-    (_, stdout, _) = process.exec_cmd([conda_path, "env", "list", "--json"])
-    env_names = [os.path.basename(env) for env in json.loads(stdout)["envs"]]
+    env_names = _list_conda_environments()
     project_env_name = _get_conda_env_name(conda_env_path, env_id)
     if project_env_name not in env_names:
         _logger.info("=== Creating conda environment %s ===", project_env_name)
-        if conda_env_path:
-            process.exec_cmd(
-                [
-                    conda_env_create_path,
-                    "env",
-                    "create",
-                    "-n",
+        try:
+            if conda_env_path:
+                process.exec_cmd(
+                    [
+                        conda_env_create_path,
+                        "env",
+                        "create",
+                        "-n",
+                        project_env_name,
+                        "--file",
+                        conda_env_path,
+                    ],
+                    stream_output=True,
+                )
+            else:
+                process.exec_cmd(
+                    [
+                        conda_env_create_path,
+                        "create",
+                        "--channel",
+                        "conda-forge",
+                        "--override-channels",
+                        "-n",
+                        project_env_name,
+                        "python",
+                    ],
+                    stream_output=True,
+                )
+        except Exception:
```

Comment:
Even if subprocess succeeds, we need to also check CondaEnvException in subprocess stdout/stderr output.

Reply:
I suggest do:
```
exit_code, stdout, stderr = process.exec_cmd(..., throw_on_error=False)
if exit_code != 0 or "CondaEnvException" in stdout or "CondaEnvException" in stderr:
  # clean created env here...
```

Reply:
> Even if subprocess succeeds, we need to also check CondaEnvException in subprocess stdout/stderr output.

Why?

Reply:
I found pip failures causes conda to fail.

Reply:
Here's a proof:

```python
% cat conda.yaml
name: yes
channels:
  - conda-forge
dependencies:
  - python=3.7.12
  - pip
  - pip:
      - mlflow==999.999.999
                                                                                                                                                                 
% conda env create -f conda.yaml
Collecting package metadata (repodata.json): done
Solving environment: done
Preparing transaction: done
Verifying transaction: done
Executing transaction: done
Ran pip subprocess with arguments:
['/Users/harutakakawamura/.pyenv/versions/miniconda3-4.7.12/envs/yes/bin/python', '-m', 'pip', 'install', '-U', '-r', '/Users/harutakakawamura/Desktop/repositories/mlflow/condaenv.j5yolst2.requirements.txt']
Pip subprocess output:

Pip subprocess error:
ERROR: Could not find a version that satisfies the requirement mlflow==999.999.999 (from versions: 0.0.1, 0.1.0, 0.2.0, 0.2.1, 0.3.0, 0.4.0, 0.4.1, 0.4.2, 0.5.0, 0.5.1, 0.5.2, 0.6.0, 0.7.0, 0.8.0, 0.8.1, 0.8.2, 0.9.0, 0.9.0.1, 0.9.1, 1.0.0, 1.1.0, 1.1.1.dev0, 1.2.0, 1.3.0, 1.4.0, 1.5.0, 1.6.0, 1.7.0, 1.7.1, 1.7.2, 1.8.0, 1.9.0, 1.9.1, 1.10.0, 1.11.0, 1.12.0, 1.12.1, 1.13, 1.13.1, 1.14.0, 1.14.1, 1.15.0, 1.16.0, 1.17.0, 1.18.0, 1.19.0, 1.20.0, 1.20.1, 1.20.2, 1.21.0, 1.22.0, 1.23.0, 1.23.1, 1.24.0)
ERROR: No matching distribution found for mlflow==999.999.999


CondaEnvException: Pip failed

                                                                                                                                                                 
% echo $?
1
```

Reply:
Please let me know if there are cases where conda succeeds when `CondaEnvException` is raised.

Reply:
I even tried installing an invalid python version (in this case, the conda env doesn't get built (obviously)) but it still returns (1) for this failure.

```
~/Documents via 🅒 base took 10s
➜ cat conda.yaml
name: hodor
channels:
  - conda-forge
dependencies:
  - python=3.14.1
  - pip
  - pip:
    - mlflow=42.42.42
(base)
~/Documents via 🅒 base
➜ conda env create -f conda.yaml
Collecting package metadata (repodata.json): done
Solving environment: failed

ResolvePackageNotFound:
  - python=3.14.1

(base)
~/Documents via 🅒 base took 9s
➜ echo $?
1
➜ conda env list
# conda environments:
#
base                  *  /Users/benwilson/opt/anaconda3
diviner-dev              /Users/benwilson/opt/anaconda3/envs/diviner-dev
mlflow-dev-env           /Users/benwilson/opt/anaconda3/envs/mlflow-dev-env
```

Reply:
Thanks Ben for the investigation :)

Reply:
Given these results, I think the current approach is ok. @WeichenXu123 WDYT?